src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/authorization/EligibleProviderMapper.kt

```diff
@@ -0,0 +1,22 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.authorization
+
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Referral
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ServiceProvider
+
+@Component
+class EligibleProviderMapper {
+  // get the service providers who are eligible to access this referral
+  fun fromReferral(referral: Referral): List<ServiceProvider> {
+    return listOf(getPrimeProvider(referral)) + getSubcontractorProviders(referral)
+  }
+
+  private fun getPrimeProvider(referral: Referral): ServiceProvider {
+    return referral.intervention.dynamicFrameworkContract.primeProvider
+  }
+
+  private fun getSubcontractorProviders(referral: Referral): List<ServiceProvider> {
```

Comment:
Is there a need to differentiate between prime and subcontractor providers at this level?

Can we have a `providersWithAccessTo(referral): List<ServiceProvider>` call that hides this complexity here? After all, the main check is "can the user's provider have access to this" -- the caller shouldn't have to check for prime, then subs, etc.

Reply:
this is a private method. intentionally private to hide this complexity.

Reply:
I completely missed that :) 👍